wain-syntax-text/src/lexer.rs

```diff
@@ -236,14 +243,70 @@ impl<'s> Lexer<'s> {
             None => return Ok(None),
         };
 
-        // Note: Content of literal will be parsed in parser so we need to traverse the content
-        // twice. It would be better to parse the content into Vec<u8> here.
-        while let Some((offset, c)) = self.chars.next() {
-            if c == '\\' {
-                self.chars.next();
-            } else if c == '"' {
-                let token = Token::String(&self.source[start + 1..offset]); // + 1 consumes '"'
-                return Ok(Some((token, start)));
+        let mut buf = vec![];
```

Comment:
> リテラルの中のパースを parser から lexer に移したのって，spec test runner と wain-syntax-text で実装を共有したかったからですよね？

それもありますが、そもそも wasm の spec でエスケープシーケンス等の解釈は 6 章の Text Format の中なので、lexer で処理するのが適切と考えたからです。
実際、spec のリポジトリにある OCaml の参照実装でも lexer で処理しています。

また、lexer.rs で
```
    String(&'source str), // Should parse the literal into Vec<u8>?
```
と
```
        // Note: Content of literal will be parsed in parser so we need to traverse the content
        // twice. It would be better to parse the content into Vec<u8> here.
```
とあったのも、lexer に移した方が良いと思った理由の一つです。
（確かに速度的にも有利かなぁ、と）

Reply:
> wasm の spec でエスケープシーケンス等の解釈は 6 章の Text Format の中なので、lexer で処理するのが適切と考えたからです。

Text Format の章は Wasm をテキストで書く際の構文を定義しているので、lex と parse 両方を定義しています。なので Text Format の章に書かれている = lex でやるべきとはならないのではと思います。

> とあったのも、lexer に移した方が良いと思った理由の一つです。

（完全に書いた記憶が残ってない…）自分で書いておいてアレですが、それを言うと float literal も2回舐めてるわけですしね…

とりあえず現状はこれで OK にして、そのあとどうするかはちょっと考えてみます。もう1つのほうのコメントご確認お願いします。

Reply:
> Text Format の章は Wasm をテキストで書く際の構文を定義しているので、lex と parse 両方の処理で必要なことを表しています。なので Text Format の章に書かれている = lex でやるべき とはならないのではと思います。

確かにその通りですね。
少なくとも 6 章全体を lexer で処理すべきではないし、そもそもできないですね。
何で自分がこの処理を lexer に書いた方がいいと思ったかと言うと、

- `string` と言う非終端記号はトークンである。
- `string` と言う非終端記号には正しいエスケープシーケンスしか含まれていない。（逆に言えば、正しくないエスケープシーケンスが含まれたものは `string` と言う非終端記号ではない）
- どうせ正しいエスケープシーケンスかどうかをチェックするのであればその場で `Vec<u8>` に変換してしまった方が二度手間を省けて楽。

だからですね。
これが 6 章に書かれてるから 6 章全体を lexer でと勘違いしましたが、全然違いましたね…

> （完全に書いた記憶が残ってない…）自分で書いておいてアレですが、それを言うと float literal も2回舐めてるわけですしね…

float literal は（も？）悩ましいですね。
そもそも `f32` と `f64` の区別が lexer の段階では付かないので…
少なくとも指数部ぐらいは `u32` に変換しちゃってもいいかなぁ、と眺めながら思ってました…

で、今文法見てて思ったんですが、これ `fnn.const` に整数渡せないですね。
エラーとするのが正しいのかも…
ちなみに wasm 1.1 では `float` と `hexfloat` の定義に小数点も指数部も必須じゃなくなったので、そもそも整数リテラルと浮動小数点リテラルの区別すら付かなくなったので渡せるようになっただけですね…
（いわゆる defect かな…）

Reply:
なるほど．正しいエスケープシーケンスが含まれてない文字列リテラルは文字列リテラルではなくなって，reserved あたりになるという話ですね．確かにどうせ舐めるならパースもしておきたいというのはそのとおりです（文字列リテラルは巨大になりがちですし）．

> 少なくとも指数部ぐらいは u32 に変換しちゃってもいいかなぁ、と眺めながら思ってました…

うーん，パースの処理はできるだけパーサ側に寄せたいですね．部分的にこっちでやって残りはあっちでやってというのはなるべく避けたいです．浮動小数点数の各パーツに分解する現状で良いのかなと思ってます．巨大になりうる文字列リテラルに対してあまりオーバーヘッドも大きく無さそうですし

> で、今文法見てて思ったんですが、これ fnn.const に整数渡せないですね。

うーん，仕様を見る限りでは渡せるようにみえます．まず

https://webassembly.github.io/spec/core/text/instructions.html

を見ると `fN.const` の引数は `fN` です．それで，`fN` の定義は符号と `fNmag` で，`fNmag` は `float` と `hexfloat` と NaN と inf です．

https://webassembly.github.io/spec/core/text/values.html#text-float

それで，`float` は

<img width="651" alt="スクリーンショット 2020-05-25 20 43 49" src="https://user-images.githubusercontent.com/823277/82810061-82016a00-9ec8-11ea-8744-b06774f46370.png">

の1つの目の branch で `.` を省略できます．また `hexfloat` は

<img width="643" alt="スクリーンショット 2020-05-25 20 44 57" src="https://user-images.githubusercontent.com/823277/82810112-9e050b80-9ec8-11ea-8977-70923a4d09b0.png">

の1つ目の branch で `.` を省略できます．これらの場合，`fN.const` の引数は整数値になっています．

Reply:
このサイトは既に 1.0 の仕様からずれてますよね…（1.1 のドラフトになってる）
W3C の Recommendation のサイトでは `.` の省略が無いです。
https://www.w3.org/TR/2019/REC-wasm-core-1-20191205/#floating-point%E2%91%A6

で、良く分からないのが、Wasm のリポジトリのタグ `wg_1.0` ではやはり `.` の省略が無いんですが、ブランチ `w3c-1.0` ではなぜか `.` の省略があるんですよね…

ちなみに、この修正は過去に Issue が上がってました。
https://github.com/WebAssembly/spec/issues/1066

仕様のバグ扱いで受け入れるのが正解ですかね…

Reply:
あっ，なるほど．そういう経緯でしたか… すみません，そこまでは把握してませんでした．貼っていただいたリンクの issue の修正を見る限り，整数も受け入れる今の実装で良さそうですね